File: packages/ridge-cli/src/key_binding.rs
//! E2EE 公钥↔身份绑定的判定核心（D-GM-10 / B3，方案 X）。
//!
//! 与桌面 `src/lib/remote/cloud/keyBinding.ts` 的 `decideKeyBinding` / `constantTimeEqual`
//! **逐字段对齐**：host/controller 在 E2EE 握手收到对端临时 X25519 公钥后，与对端经
//! **已认证信令**旁路（A 层 `e2ee-pubkey`，**仅 eph_pub、不带 sig** —— 方案 X，relay 零
//! 密码学材料）转发回来的公钥做恒定时间比对。不一致 = relay 层主动 MITM（给两端各发攻击
//! 者自己的公钥）→ 拒绝会话。
//!
//! 启用门 = **信令公钥到达性**（而非 D9 `$/hello` 能力位）：`$/hello` 在 E2EE 握手完成后
//! 才发，太晚无法据它在握手时拒绝。改用更稳健的"信令公钥到达性"——DataChannel 网络中间人
//! 无法篡改另一条已认证 TLS 信令上转发的公钥，故"收到信令公钥"即强制比对；"宽限期内未收
//! 到"暂缓；"宽限期过仍未收到"判对端为不发该信令的旧端 → 回落 relay-trust（向后兼容，
//! 旧端不回归）。DataChannel MITM 无法借此回落逃逸——它无法阻止信令公钥经独立 TLS 到达。
//!
//! 纯模块（无 I/O、无网络），可独立单测；session.rs 据其判定结果决定 accept / 断开 / 等待。

use crate::e2ee::PUB_KEY_LEN;

/// 恒定时间比较两个字节序列是否相等（防计时侧信道）。长度不同立即返回 false
/// （长度本身非秘密）；等长时逐字节累积差异，不提前退出。
pub fn constant_time_equal(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for i in 0..a.len() {
        diff |= a[i] ^ b[i];
    }
    diff == 0
}

/// 三态判定（与 keyBinding.ts `KeyBindingDecision` 对齐）。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum KeyBindingDecision {
    /// 校验通过（或宽限期回落 relay-trust）：可放行业务。
    Accept,
    /// 检测到 MITM：握手公钥 ≠ 信令旁路公钥 → 必须断开。
    Reject,
    /// 信令旁路公钥尚未到达且宽限期未过 → 暂缓决定（既不放行也不拒绝）。
    Wait,
}

/// 一个连接**已决**的 B3 绑定模式（诊断/日志可读）。对应桌面 `keyBinding.ts`
/// `KeyBindingMode` 的两个**终态**；其 `pending`（未决）态在 cli 侧由
/// `binding_decided: bool` 跟踪，无需独立枚举值。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum KeyBindingMode {
    /// 收到信令旁路公钥并比对一致（严格绑定生效）。
    Enforced,
    /// 宽限期内未收到信令公钥 → 回落 v1（对端疑为旧端）。
    RelayTrust,
}

/// B3 在线判定（与 keyBinding.ts::decideKeyBinding 逐字段对齐）。
///
/// - `handshake_pub`：E2EE DataChannel 握手帧解出的对端公钥（`0x01||pub32`）。
/// - `signaling_pub`：对端经已认证信令转发回来的公钥；尚未到达为 `None`。
/// - `grace_expired`：宽限期是否已过（仍未收到 `signaling_pub` 时用于回落判定）。
pub fn decide_key_binding(
    handshake_pub: &[u8],
    signaling_pub: Option<&[u8]>,
    grace_expired: bool,
) -> KeyBindingDecision {
    if let Some(sig) = signaling_pub {
        if sig.len() != PUB_KEY_LEN || handshake_pub.len() != PUB_KEY_LEN {
            return KeyBindingDecision::Reject; // 防御性：任一公钥长度非法 → 拒绝
        }
        return if constant_time_equal(handshake_pub, sig) {
            KeyBindingDecision::Accept
        } else {
            KeyBindingDecision::Reject // 握手公钥 ≠ 信令旁路公钥 → MITM
        };
    }
    // 尚无信令公钥：宽限期内等待；过期则回落 relay-trust（对端疑为旧端，不发信令公钥）。
    if grace_expired {
        KeyBindingDecision::Accept
    } else {
        KeyBindingDecision::Wait
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn k(b: u8) -> [u8; PUB_KEY_LEN] {
        [b; PUB_KEY_LEN]
    }

    #[test]
    fn constant_time_equal_basic() {
        assert!(constant_time_equal(&k(7), &k(7)));
        assert!(!constant_time_equal(&k(7), &k(8)));
        assert!(!constant_time_equal(&[1, 2, 3], &[1, 2])); // 长度不同立即 false
    }

    #[test]
    fn accept_when_signaling_matches_handshake() {
        // 信令公钥 == 握手公钥 → enforced accept。
        assert_eq!(
            decide_key_binding(&k(1), Some(&k(1)), false),
            KeyBindingDecision::Accept
        );
    }

    #[test]
    fn reject_when_signaling_differs() {
        // 信令公钥 ≠ 握手公钥 → MITM → reject。
        assert_eq!(
            decide_key_binding(&k(1), Some(&k(2)), false),
            KeyBindingDecision::Reject
        );
    }

    #[test]
    fn reject_on_bad_length_signaling() {
        assert_eq!(
            decide_key_binding(&k(1), Some(&[1, 2, 3]), false),
            KeyBindingDecision::Reject
        );
    }

    #[test]
    fn wait_before_grace_when_no_signaling() {
        // 宽限期未过、信令公钥未到 → 暂缓。
        assert_eq!(
            decide_key_binding(&k(1), None, false),
            KeyBindingDecision::Wait
        );
    }

    #[test]
    fn accept_after_grace_without_signaling_relay_trust() {
        // 宽限期过仍无信令公钥 → 回落 relay-trust（accept，兼容旧端）。
        assert_eq!(
            decide_key_binding(&k(1), None, true),
            KeyBindingDecision::Accept
        );
    }
}

File: packages/remote/src/shared/cloud/keyBinding.ts
// Ridge Cloud — E2EE 公钥↔身份绑定的**校验核心**（D-GM-10 / B3）。
//
// 设计见 docs/plans/d-gm-10-e2ee-key-binding-design.md。这是该安全特性里**可在
// wind 单仓库自主实现并单测**的一半：host/controller 在 E2EE 握手收到对端临时
// 公钥后，用本模块判定「该公钥是否 == cloud 经**已认证信令**旁路转发回来的对端
// 公钥」。不一致 = 检测到 relay 层主动 MITM（给两端各发了攻击者自己的公钥），
// 拒绝会话。
//
// 威胁模型（与设计 §2 一致）：把"任何能转发 E2EE 握手的中间人都能 MITM"收窄到
// "须同时攻陷已认证 TLS 信令 + 持双方 deviceJWT 认证态"。完全攻陷 cloud auth 后端
// 不在防护内（既定边界）。
//
// 本模块**不**自己取"对端信令公钥"——那来自 ridge-cloud 的信令转发（`e2ee-peer-
// pubkey`）。本模块只产出 B3 在线判定所需的**纯函数**（decideKeyBinding），
// 由双 provider 在握手/信令公钥到达时调用。
//
// 兼容（D9 能力协商）：仅当双方都公告 `e2ee-bind` 能力时才启用严格比对
// （`enabled=true`）；否则回落 relay-trust v1（`enabled=false` → 放行），避免老
// controller 回归。

import { PUBKEY_LEN } from './e2ee';

/**
 * 恒定时间比较两个字节数组是否相等（防计时侧信道）。长度不同立即返回 false
 * （长度本身非秘密）。等长时不提前退出，逐字节累积差异。
 */
export function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff === 0;
}

// makeKeyBindingVerifier / KeyBindingOptions 已随 §5.5 桥钩子退役删除（S1-F5，
// iteration 8 G4）：其目标场景由下方 decideKeyBinding（B3 在线判定）+ 0x02 帧覆盖。

/**
 * 一个连接的 B3 绑定模式（诊断/测试可读）。
 * - `pending`：尚未判定。
 * - `enforced`：收到信令旁路公钥并比对一致（严格绑定生效）。
 * - `relay-trust`：宽限期内未收到信令公钥 → 回落 v1（对端疑为旧端）。
 */
export type KeyBindingMode = 'pending' | 'enforced' | 'relay-trust';

/** {@link decideKeyBinding} 的三态判定。 */
export type KeyBindingDecision =
  | 'accept' // 校验通过（或回落 relay-trust）：可标记 connected
  | 'reject' // 检测到 MITM：握手公钥 ≠ 信令旁路公钥 → 必须断开
  | 'wait'; //  信令旁路公钥尚未到达且宽限期未过 → 暂缓决定（既不放行也不拒绝）

/**
 * B3 在线判定：用「**信令旁路确认是否到达**」作启用门，而非 D9 `$/hello` 能力位。
 *
 * 为何不用 `$/hello`：`$/hello` 在 E2EE 握手**完成后**才由 L2 发出（见 rpcClient/
 * bridge），那时连接已 `connected`，太晚——无法在握手时据它决定是否拒绝。改用更稳健
 * 的**信令公钥到达性**：DataChannel 网络中间人**无法**篡改另一条已认证 TLS 信令上转发
 * 的公钥，故"收到了信令公钥"即可强制比对；"宽限期内未收到"暂缓；"宽限期过仍未收到"判
 * 定对端为不发信令公钥的旧端 → 回落 relay-trust（向后兼容）。DataChannel MITM 无法
 * 借此回落逃逸——它无法阻止信令公钥经独立 TLS 通道到达。
 *
 * @param handshakePub   E2EE DataChannel 握手帧里的对端公钥（0x01||pub32 解出）。
 * @param signalingPub   cloud 经已认证信令转发回来的对端公钥；尚未到达为 null。
 * @param graceExpired   宽限期是否已过（仍未收到 signalingPub 时用于回落判定）。
 */
export function decideKeyBinding(
  handshakePub: Uint8Array,
  signalingPub: Uint8Array | null,
  graceExpired: boolean,
): KeyBindingDecision {
  if (signalingPub) {
    if (signalingPub.length !== PUBKEY_LEN || handshakePub.length !== PUBKEY_LEN) {
      return 'reject'; // 防御性：任一公钥长度非法 → 拒绝
    }
    return constantTimeEqual(handshakePub, signalingPub) ? 'accept' : 'reject';
  }
  // 尚无信令公钥：宽限期内等待；过期则回落 relay-trust（对端疑为旧端，不发信令公钥）。
  return graceExpired ? 'accept' : 'wait';
}
